test.sh

```diff
@@ -21,25 +21,118 @@ TESTDIRS="analysis \
           # cmd
           # Godeps
 
+# We need to know, for github-pr-status, what the triggering commit is.
+# Assume first it's the travis commit (for builds of master), unless we're
+# a PR, when it's actually the first parent.
+TRIGGER_COMMIT=${TRAVIS_COMMIT}
+if [ "${TRAVIS_PULL_REQUEST}" != "false" ] ; then
+  revs=$(git rev-list --parents -n 1 HEAD)
+  # The trigger commit is the last ID in the space-delimited rev-list
+  TRIGGER_COMMIT=${revs##* }
+fi
+
+start_context() {
+  CONTEXT="$1"
+  echo "Starting [${CONTEXT}]"
+}
+
+end_context() {
+  CONTEXT=""
+}
+
+update_status() {
+  if [ "${TRAVIS}" == "true" ] && [ "x${CONTEXT}" != "x" ] ; then
+    github-pr-status --authfile "$(pwd)/test/github-secret.json" \
+      --owner "letsencrypt" --repo "boulder" \
+      status --sha "${TRIGGER_COMMIT}" --context "${CONTEXT}" $*
+  fi
+}
+
 run() {
   echo "$*"
-  if $*; then
+  $* 2>&1
+  local status=$?
+
+  if [ ${status} -eq 0 ]; then
+    update_status --state success
     echo "success: $*"
   else
     FAILURE=1
+    update_status --state failure
     echo "failure: $*"
   fi
+
+  return ${status}
 }
 
+run_and_comment() {
+  if [ "x${TRAVIS}" = "x" ] || [ "${TRAVIS_PULL_REQUEST}" == "false" ] ; then
+    run $*
+  else
+    result=$(run $*)
+    local status=$?
+    # Only send a comment if exit code > 0
+    if [ ${status} -ne 0 ] ; then
+      echo ${result} | github-pr-status --authfile "$(pwd)/test/github-secret.json" \
```

Comment:
In order to increase readability of the problematic section of the log it might be nice to wrap `${result}` with muli-line code tags (``` `\`\`\`\\n${result}\\n\`\`\` ```)